Basic spatial module docs (#162)
EtherealEngine/etherealengine-docs: docs/manual/03_modules/02_spatial/01_transforms.md

# Turning entities into spatialized objects

## Transforms

> The first concept to spatialize our entities is to introduce the `TransformComponent`. This stores the spatial information for our entity, it's `position`, `rotation` and `scale`. These property are bitecs proxified three.js math objects: a Vector3 for position, Quaternion for rotation another Vector3 for scale.

Reviewer: ```suggestion
The first concept to spatialize our entities is to introduce the `TransformComponent`. This stores the spatial information for our entity, its `position`, `rotation`, and `scale`. These properties are bitecs-proxified three.js math objects: a Vector3 for position, a Quaternion for rotation, and another Vector3 for scale.
```

> There are two more properties, the `matrix` and `matrixWorld`, which are threejs Matrix4 properties. `matrix` stores this transform relative to it's parent, and `matrixWorld` stores it relative to the absolute origin.

Reviewer: ```suggestion
There are two more properties, the `matrix` and `matrixWorld`, which are three.js Matrix4 properties. `matrix` stores this transform relative to its parent, and `matrixWorld` stores it relative to the absolute origin.
```

## Hierarchy

> Our next component is the `EntityTreeComponent` which stores the parent/child relationships: a `parentEntity` entity and `children` array of entities. Changing either of these via setComponent will automatically update the parent and children entities to reflect this change.

Reviewer: ```suggestion
Our next component is the `EntityTreeComponent` which stores the parent/child relationships: a `parentEntity` and `children` array of entities. Changing either of these via `setComponent` will automatically update the parent and children entities to reflect this change. 
```

> To optimize upon traditional hierarchy based calculations, the `TransformSystem` automatically sorts an complete list of all spatial entities in order to utilize `bitecs'` SoA CPU caching to allow the javascript engine to calculate matrices as fast as possible.

Reviewer: ```suggestion
To optimize traditional hierarchy-based calculations, the `TransformSystem` automatically sorts a complete list of all spatial entities in order to utilize `bitecs`’ SoA CPU caching to allow the JavaScript engine to calculate matrices as fast as possible.
```

## Physics

Physics is split into two systems.

> The first is the `PhysicsSystem` which updates the ECS and physics bodies before & after simulation, as well running the simulation. Substeps can be configured via `PhysicsState`, and the simulation timestep can be configured via `ECSState`. Beyond this, the system has a few neat tricks, such as interpolating kinematic rigidbodies between substeps, and automatically populating & depopulating collision events into the `CollisionComponent`. 

Reviewer: ```suggestion
The first is the `PhysicsSystem` which updates the ECS and physics bodies before and after simulation, as well as running the simulation. Substeps can be configured via `PhysicsState`, and the simulation timestep can be configured via `ECSState`. Beyond this, the system has a few neat tricks, such as interpolating kinematic rigidbodies between substeps, and automatically populating and depopulating collision events into the `CollisionComponent`.
```
